### DISCUSSION TITLE
Inferring stereochemistry

### DISCUSSION THREAD

**Discussion by orionarcher** [openforcefield/discussions#20] [General]
Is there any way in Openff toolkit to infer stereochemistry from molecule coordinates?

If not, are there other tools the devs would recommend that can do this?

Ideally I'd be able to infer stereo labels like E/Z, cis-trans directly from the molecular graph + coordinates, which I am extracting from DFT.

**Comment by j-wags** [comment]
Not directly in the OpenFF toolkit - We have to be careful not to cross the line into cheminformatics-land, since that's a huge behavior surface area to maintain.

What I'd recommend here is taking a quick round trip with `to_rdkit` and `from_rdkit`. I'll put together a code snippet when I have a minute, but roughly something like [this](https://github.com/openforcefield/openff-toolkit/blob/b52879569a0344878c40248ceb3bd0f90348076a/openff/toolkit/utils/rdkit_wrapper.py#L279-L290), where the important part is telling RDKit to assign stereo from 3D.

**Comment by j-wags** [comment]
Here you go:

```python
from openff.toolkit import Molecule
from rdkit import Chem

## Make a graph mol with stereo but no conformers

mol = Molecule.from_smiles('[C@H](F)(Cl)\C=C/C')
mol.visualize(backend="rdkit")
```
<img width="405" alt="Screen Shot 2023-04-20 at 1 56 04 PM" src="https://user-images.githubusercontent.com/16329175/233485823-ce73df61-551e-43e0-8cd1-dd4c27775b2b.png">

```python
# Generate a conformer with the correct stereo and make a copy of it
mol.generate_conformers()
conf = mol.conformers[0]
mol.visualize(backend='nglview')
```

<img width="526" alt="Screen Shot 2023-04-20 at 1 56 34 PM" src="https://user-images.githubusercontent.com/16329175/233485918-4502e62d-e683-432d-8644-6f73f8641b3b.png">


```python
# Remove the molecule's stereo (sorry, this is hacky but I'm just doing this to get to the starting point of your use case)
mol._invalidate_cached_properties()
for atom in mol.atoms:
    atom.stereochemistry = None
for bond in mol.bonds:
    bond._stereochemistry = None
    
# This removes the molecule's stereo information (and conformers)
mol.visualize(backend="rdkit")
```
<img width="511" alt="Screen Shot 2023-04-20 at 1 56 58 PM" src="https://user-images.githubusercontent.com/16329175/233485975-9c5c56b4-d0c6-4b9e-bd20-e3af0a5bea83.png">


```python
# To verify - Trying to roundtrip this mol leads to an undefinedstereoerror since there's no graph stereo any more
Molecule.from_rdkit(mol.to_rdkit())
```

> UndefinedStereochemistryError: Unable to make OFFMol from RDMol: RDMol has unspecified stereochemistry. RDMol name: 
> Undefined chiral centers are:
>  - Atom C (index 0)
> Bonds with undefined stereochemistry are:
>  - Bond 4 (atoms 4-5 of element (C-C)

```python
# Add the conformer with the correct stereo back, but note that even though the stereo is present in the 3D, OpenFF and RDKit haven't updated the graph representation.
mol.add_conformer(conf)
mol.visualize(backend="rdkit")
```
<img width="495" alt="Screen Shot 2023-04-20 at 1 58 00 PM" src="https://user-images.githubusercontent.com/16329175/233486177-dd69beeb-4906-484c-bd2f-3f68aeaef5b9.png">

```python
# This should be the part you want - We send the molecule to RDKit and tell it to assign stereo from 3D
rdmol = mol.to_rdkit()
Chem.AssignStereochemistryFrom3D(rdmol)
mol2 = Molecule.from_rdkit(rdmol)

# Now the graph stereo has been added back based on the 3d coords :-)
mol2.visualize(backend="rdkit")
```
<img width="404" alt="Screen Shot 2023-04-20 at 1 58 18 PM" src="https://user-images.githubusercontent.com/16329175/233486247-54316f22-9c6d-49b2-a08b-0cd77e50b257.png">

**Comment by orionarcher** [comment]
Awesome! Thank you so much @j-wags!
